### ISSUE TITLE
Support negative Version numbers for time travle

### ISSUE THREAD

**Issue by gbrueckl** [delta-io/delta#636]
I often find myself in situation where I want to compare the current state of a delta table with the previous state (e.g. before a MERGE/UPDATE)
right now this is always a bit cumbersome as I need to get the version/timestamp of the previous transaction using `DESCRIBE HISTORY` before I can use `[VERSION / TIMESTAMP] AS OF xxx`

It would be great to have the ability to define negative version numbers like `VERSION AS OF -1` to get the previous version, `VERSION AS OF -2` to get the version two commits ago etc.
basically similar to what you can do in Python with arrays to select the trailing elements

**Comment by SpaceRangerWes** [comment]
@zsxwing I can take this if not already in progress

**Comment by SpaceRangerWes** [comment]
@zsxwing should this be supported in the `startingVersion` for streaming scenarios? I was thinking of leaving that out because it would be idempotent unlike a true version number.

Example: If I set `startingVersion=-1` on run 1 and have 123 versions, I'll start at version 122. If I run the code again on run 2 some time later, I may start at version 128.

**Comment by zsxwing** [comment]
> Example: If I set `startingVersion=-1` on run 1 and have 123 versions, I'll start at version 122. If I run the code again on run 2 some time later, I may start at version 128.

I think this applies to batch queries and streaming queries using `startingVersion=latest` as well. For streaming queries, once we start the query, it should be written to the checkpoint and after that running the query will ignore `startingVersion`. Hence it should be fine.

**Comment by rahulsmahadev** [comment]
@gbrueckl @SpaceRangerWes After thinking about this a little more I'm a little worried that this can be misused and can lead to confusing behavior for users. Can you help me understand your use case better ?

If it is to just find what rows were added/updated by your command just finding the last version using history and selecting that version would work (although its two steps instead of 1).

The problem with supporting `-1` is that It can be misunderstood by users in a way that they could think it's their last commit, but it could be anyone's commit on the table. Imagine a periodic optimize job took place between a user's last 2 commits. Apart from `-1` I do not see much value for `-n`.

Unless I'm missing the use case I'm inclined to decline this PR #673, feel free to point out some other use cases if I'm missing them.

**Comment by SpaceRangerWes** [comment]
@rahulsmahadev your explanation is good enough for me. Can't really argue with it.

**Comment by gbrueckl** [comment]
@rahulsmahadev thanks for the explanation - that makes a lot of sense
the initial idea was to have an easy/generic way to compare the existing version with the previous version - mainly for data validation purposes
in pure Spark SQL it is hard to get this information otherwise (PySpark, Scala, etc. would of course still work just fine as you are much more flexible there)
maybe another approach would be to allow sub-queries over `DESCRIBE HISTORY` and the ability to make `VERSION AS OF x` also able to use sub-queries?

**Comment by rahulsmahadev** [comment]
Thanks, will close this issue for now.
